### PR TITLE
Fix incorrectly named "ip-mask" and "api-cors-headers" options

### LINKED ISSUE #22807: [1.11.1] Use `ip-masq` everywhere instead of both `ip-mask` and `ip-masq`.
The Docker Engine daemon CLI param is `--ip-masq`, so do not use `ip-mask` parameter instead in `daemon.json`.

### PR DESCRIPTION
**Pull request by coldTea214**
**\- What I did**
Fixed small docs error

**\- Description for the changelog**
Fix "ip-mask" and "api-cors-headers" configuration not working. These options were named incorrectly and have been renamed to "ip-masq" and "api-cors-header" to match the daemon options with the same name.

fixes https://github.com/docker/docker/issues/22807

Signed-off-by: Wang Xing hzwangxing@corp.netease.com

### CODE REVIEW

File: docs/reference/commandline/dockerd.md [hunk 1]
```diff
@@ -930,7 +930,7 @@ This is a full example of the allowed configuration options in the file:
 	"ipv6": false,
 	"iptables": false,
 	"ip-forward": false,
-	"ip-mask": false,
+	"ip-masq": false,
```
Comment (thaJeztah) [inline docs/reference/commandline/dockerd.md]: Hm, I think this needs more than a documentation change; and we may need to accept both; see https://github.com/docker/docker/blob/28b7dc51aec6a237af187badfc1953eb9dbbb142/daemon/config_unix.go#L44

File: docs/reference/commandline/dockerd.md [hunk 2]
```diff
@@ -921,7 +921,7 @@ This is a full example of the allowed configuration options in the file:
 	"tlscacert": "",
 	"tlscert": "",
 	"tlskey": "",
-	"api-cors-headers": "",
+	"api-cors-header": "",
```
Comment (thaJeztah) [inline docs/reference/commandline/dockerd.md]: Same for this one; we should probably add a fallback for the "wrong" one; https://github.com/docker/docker/blob/e9117578a7837eda31d690d5e51d09d43a6a8fca/daemon/config.go#L91

**Comment by thaJeztah** [comment]
Thanks @wangxing1517 these look like typos indeed, and we probably want to fix them, but that would require code changes as well. We should discuss if we need a "fallback", or just describe it in the changelog

ping @calavera @yongtang wdyt?

**Comment by yongtang** [comment]
@thaJeztah @wangxing1517 I think if we want to play it safe and not causing big issues on user side, we probably need to update the code so that `ip-mask/ip-masq` and `api-cors-headers/api-cors-header` could co-exist for some time.

Maybe we could deprecate `ip-mask` and `api-cors-headers` starting from 1.12. Then we remove them after 1.14?

**Comment by yongtang** [comment]
I can try to create a pull request if this is the change we want to make.

**Comment by thaJeztah** [comment]
I'm +1 for making those changes.

one thing I'm wondering is, if "api-cors-header" should actually be an option that's allowed to be specified multiple times (i.e. `--api-cors-header=[]`). Looks like we currently document using a [comma-separated list of headers](https://github.com/docker/docker/blob/e9117578a7837eda31d690d5e51d09d43a6a8fca/man/dockerd.8.md#options); I actually think that example is actually invalid as that list may need to be [space-separated list](http://www.w3.org/TR/cors/#access-control-allow-origin-response-header). 

> In practice the origin-list-or-null production is more constrained. Rather than allowing a space-separated list of origins, it is either a single origin or the string "null".

So, if multiple hosts are allowed, the recommended approaches to check the received host-header against a whitelist, and return that host in the `Access-Control-Allow-Origin` header, if it's whitelisted.

Actually, I think we should set a `Vary` header as well if `api-cors-header` is set [from developer.mozilla.org](https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS#Access-Control-Allow-Origin);

> If the server specifies an origin host rather than "*", then it must also include Origin in the Vary response header to indicate to clients that server responses will differ based on the value of the Origin request header.

Looks like we [currently don't do that](https://github.com/docker/docker/blob/9438572fc00f928ec83c884992d5492a7a68babb/api/server/middleware/cors.go#L22-L37)

**Comment by thaJeztah** [comment]
ping @dqminh perhaps you have some thoughts on those headers? I saw you're at cloudflare now, so you probably deal with those things more frequently :smile:

**Comment by thaJeztah** [comment]
@wangxing1517 looks like `ip-mask` currently doesn't actually work, because the validation function checks for the actual *flag* name, see https://github.com/docker/docker/issues/22807#issuecomment-220288664.

Can you:

- change this line [daemon/config_unix.go#L44](https://github.com/docker/docker/blob/28b7dc51aec6a237af187badfc1953eb9dbbb142/daemon/config_unix.go#L44) to also use `ip-masq`
- change this line [daemon/config.go#L91](https://github.com/docker/docker/blob/e9117578a7837eda31d690d5e51d09d43a6a8fca/daemon/config.go#L91) to also use `api-cors-header`

Then we should be able to merge this

**Comment by dqminh** [comment]
> So, if multiple hosts are allowed, the recommended approaches to check the received host-header against a whitelist, and return that host in the Access-Control-Allow-Origin header, if it's whitelisted.

Yes i think this is correct. If i understand correctly, the current code returns a comma-separated list which is wrong. Also we should add `Vary: Origin` header too.

**Comment by thaJeztah** [comment]
Thanks @dqminh! I'll open an issue for it, and possibly look into opening a PR if I have time

**Comment by thaJeztah** [comment]
oh, looks like this was updated

LGTM, thanks @wangxing1517 

ping @vdemeester PTAL

**Comment by vdemeester** [comment]
LGTM :tiger:

**Pull request merged**